src/server.js

```diff
@@ -0,0 +1,211 @@
+/**
+ * Server to Intercept the client's requests and handle them on their behalf.
+ * Initiates stats and connectivity checks when a requests fails.
+ * It also responds in selenium understandable error when a request fails
+ * at tool.
+ */
+
+var http = require('http');
+var url = require('url');
+var uuidv4 = require('uuid/v4');
+var Utils = require('./utils');
+var constants = require('../config/constants');
+var ReqLib = require('./requestLib');
+var RdGlobalConfig = constants.RdGlobalConfig;
+
+var RdHandler = {
+
+  _requestCounter: 0,
+
+  /**
+   * Generates the request options template for firing requests based on
+   * whether the user had provided any proxy input or not.
+   */
+  generatorForRequestOptionsObject: function () {
+    RdHandler._reqObjTemplate = {
+      method: null,
+      headers: {},
+      host: null,
+      port: null,
+      path: null
+    };
+
+    if (RdGlobalConfig.proxy) {
+      RdHandler._reqObjTemplate.host = RdGlobalConfig.proxy.host;
+      RdHandler._reqObjTemplate.port = RdGlobalConfig.proxy.port;
+      
+      if (RdGlobalConfig.proxy.username && RdGlobalConfig.proxy.password) {
+        RdHandler._reqObjTemplate.headers['Proxy-Authorization'] = Utils.proxyAuthToBase64(RdGlobalConfig.proxy);
+      }
+
+      /**
+       * Sets the internal method to generate request options if external/upstream
+       * proxy exists
+       * @param {http.IncomingMessage} clientRequest 
+       * @returns {Object}
+       */
+      RdHandler._generateRequestOptions = function (clientRequest) {
+        var parsedClientUrl = url.parse(clientRequest.url);
+        var headersCopy = Object.assign({}, clientRequest.headers, RdHandler._reqObjTemplate.headers);
+        var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
+        requestOptions.path = parsedClientUrl.href;
+        requestOptions.method = clientRequest.method;
+        requestOptions.headers = headersCopy;
+        return requestOptions;
+      };
+    } else {
+
+      /**
+       * Sets the internal method to generate request options if external/upstream
+       * doesn't exists
+       * @param {http.IncomingMessage} clientRequest 
+       * @returns {Object}
+       */
+      RdHandler._generateRequestOptions = function (clientRequest) {
+        var parsedClientUrl = url.parse(clientRequest.url);
+        var requestOptions = Object.assign({}, RdHandler._reqObjTemplate);
+        requestOptions.host = parsedClientUrl.hostname;
+        requestOptions.port = parsedClientUrl.port || 80;
+        requestOptions.path = parsedClientUrl.path;
+        requestOptions.method = clientRequest.method;
+        requestOptions.headers = clientRequest.headers;
+        if (parsedClientUrl.auth) {
+          requestOptions.headers['authorization'] = Utils.proxyAuthToBase64(parsedClientUrl.auth);
+        }
```

Comment:
Why need of this when we can copy the headers from client, ideally that should contain the basicAuth creds.

Can we also add the functionality to parse the URL and if basicAuth creds present, insert them into headers.

Reply:
I think what you are referring to is what is happening at that line. The request that comes in can contain the auth info in headers or in the url itself. Thus, parsing it at line 65 if it exists in the URL. If yes, I convert it into base64 and add it manually in the headers. If its already there in the headers, it will be handled at the line 71.
Or am i missing something?

Reply:
Cool makes sense 👍